crypto/mlkem/mlkem-variant.c

```diff
@@ -0,0 +1,1215 @@
+ * modular reduction) the candidate values {0xdab, 0xefc} for d1, d2.
+ */
+static int scalar_from_keccak_vartime(scalar *out, EVP_MD_CTX *mdctx)
+{
+    int done = 0;
+    uint8_t block[180];
+    size_t i;
+    uint16_t d1, d2;
+
+    while (done < DEGREE) {
+        if (!EVP_DigestSqueeze(mdctx, block, sizeof(block)))
+            return 0;
+        for (i = 0; i < sizeof(block) && done < DEGREE; i += 3) {
+            d1 = block[i] + ((block[i + 1] & 0x0f) << 8);
+            d2 = (block[i + 1] >> 4) + (block[i + 2] << 4);
+            if (d1 < kPrime)
+                out->c[done++] = d1;
+            if (d2 < kPrime && done < DEGREE)
+                out->c[done++] = d2;
+        }
+    }
+    return 1;
+}
+
+typedef const uint8_t augmented_seed_t[MLKEM_RANDOM_BYTES + 1];
+typedef int cbd_func(scalar *, augmented_seed_t, ossl_mlkem_ctx *);
+
+# if ETA1 == 2 || ETA2 == 2
+/*
+ * Algorithm 7 from the spec, with eta fixed to two and the PRF call
+ * included. Creates binominally distributed elements by sampling 2*|eta| bits,
+ * and setting the coefficient to the count of the first bits minus the count of
+ * the second bits, resulting in a centered binomial distribution. Since eta is
+ * two this gives -2/2 with a probability of 1/16, -1/1 with probability 1/4,
+ * and 0 with probability 3/8.
+ */
+static
+int scalar_cbd_2(scalar *out, augmented_seed_t input, ossl_mlkem_ctx *mlkem_ctx)
+{
+    uint8_t randbuf[2 * 2 * DEGREE / 8];    /* 64 * eta */
+    int i;
+    uint8_t byte;
+    uint16_t value;
+
+    if (!prf(randbuf, sizeof(randbuf), input, mlkem_ctx))
+        return 0;
+    for (i = 0; i < DEGREE; i += 2) {
+        byte = randbuf[i / 2];
+        value = kPrime;
+        value += (byte & 1) + ((byte >> 1) & 1);
+        value -= ((byte >> 2) & 1) + ((byte >> 3) & 1);
+        out->c[i] = reduce_once(value);
+        byte >>= 4;
+        value = kPrime;
+        value += (byte & 1) + ((byte >> 1) & 1);
+        value -= ((byte >> 2) & 1) + ((byte >> 3) & 1);
+        out->c[i + 1] = reduce_once(value);
+    }
+    return 1;
+}
+#  if ETA1 == 2
+#   define cbd_eta1_with_prf scalar_cbd_2
+#  endif
+#  if ETA2 == 2
+#   define cbd_eta2_with_prf scalar_cbd_2
+#  endif
+# endif
+
+# if ETA1 == 3 || ETA2 == 3
+/*
+ * Algorithm 7 from the spec, with eta fixed to two and the PRF call
+ * included. Creates binominally distributed elements by sampling 3*|eta| bits,
+ * and setting the coefficient to the count of the first bits minus the count of
+ * the second bits, resulting in a centered binomial distribution.
+ */
+static
+int scalar_cbd_3(scalar *out, augmented_seed_t input, ossl_mlkem_ctx *mlkem_ctx)
+{
+    uint8_t randbuf[6 * DEGREE / 8];    /* 64 * eta */
+    int i = 0, j = 0;
+    uint8_t byte;
+    uint16_t value;
+
+    if (!prf(randbuf, sizeof(randbuf), input, mlkem_ctx))
+        return 0;
+    /* Unrolled loop uses 3 bytes at a time, yielding 4 values (6 bits each) */
+    while (j < (int) sizeof(randbuf)) {
```

Comment:
Is it worth conditioning the extra two iterations on `OPENSSL_SMALL_FOOTPRINT`?

Reply:
I think you've misunderstood the logic here. Unclear what you mean by "extra two iterations". This loop consumes a fixed amount of data, 6 bits per generated `R_q` "scalar".